lazy-pages/src/sys/unix.rs

```diff
@@ -109,7 +109,102 @@ where
+    #[display(fmt = "Cannot get information about old signal stack: {}", _0)]
+    OldStack(Errno),
+    #[display(fmt = "Cannot mmap space for signal stack: {}", _0)]
+    Mmap(Errno),
+    #[display(fmt = "Cannot set new signal stack: {}", _0)]
+    SigAltStack(Errno),
+}
+
+fn init_for_thread_internal() -> Result<(), ThreadInitError> {
+    use core::{mem, ptr};
+
+    // Should be enough for lazy-pages signal handler.
+    // Equal to libc::SIGSTKSZ on macos M1.
+    const SIGNAL_STACK_SIZE: usize = 0x20000;
+
+    enum StackInfo {
+        UseOldStack,
+        NewStack(*mut libc::c_void),
+    }
+
+    impl Drop for StackInfo {
+        fn drop(&mut self) {
+            if let StackInfo::NewStack(mmap_ptr) = self {
+                unsafe {
+                    // Deallocate the stack memory.
+                    if libc::munmap(*mmap_ptr, SIGNAL_STACK_SIZE) != 0 {
+                        log::error!(
+                            "Cannot deallocate signal stack memory during the thread shutdown: {}",
+                            errno::errno()
+                        );
+                    }
+                }
+            }
+        }
+    }
+
+    unsafe fn init_sigstack() -> Result<StackInfo, ThreadInitError> {
+        // Check whether old signal stack exist and suitable for lazy-pages signal handler.
+        let mut old_stack = mem::zeroed();
+        let res = libc::sigaltstack(ptr::null(), &mut old_stack);
+        if res != 0 {
+            return Err(ThreadInitError::OldStack(errno::errno()));
+        }
+        if old_stack.ss_flags & libc::SS_DISABLE == 0 && old_stack.ss_size >= SIGNAL_STACK_SIZE {
+            return Ok(StackInfo::UseOldStack);
+        }
+
+        // Alloc memory for new signal stack.
+        let ptr = libc::mmap(
+            ptr::null_mut(),
+            SIGNAL_STACK_SIZE,
+            libc::PROT_READ | libc::PROT_WRITE,
+            libc::MAP_PRIVATE | libc::MAP_ANON,
+            -1,
+            0,
+        );
+        if ptr == libc::MAP_FAILED {
+            return Err(ThreadInitError::Mmap(errno::errno()));
+        }
+
+        // Mark allocated memory as new signal stack.
+        let new_stack = libc::stack_t {
+            ss_sp: ptr,
+            ss_flags: 0,
+            ss_size: SIGNAL_STACK_SIZE,
+        };
+        let res = libc::sigaltstack(&new_stack, ptr::null_mut());
+        if res != 0 {
+            return Err(ThreadInitError::SigAltStack(errno::errno()));
+        }
+
+        log::debug!(
+            "Set new signal stack: ptr = {:?}, size = {:#x}",
+            ptr,
+            SIGNAL_STACK_SIZE
+        );
+
+        Ok(StackInfo::NewStack(ptr))
+    }
+
+    thread_local! {
+        static TLS: Result<StackInfo, ThreadInitError> = unsafe { init_sigstack() };
+    }
+
+    TLS.with(|tls| tls.as_ref().map(|_| ()).map_err(|err| *err))
+}
+
+pub(crate) unsafe fn init_for_thread() -> Result<(), String> {
+    init_for_thread_internal().map_err(|err| err.to_string())
```

Comment:
We have no need to convert error into string

Reply:
Ok, but I cannot use Errno outside. Of course I can use io::Error, but what the difference between using io::Error and String in our case ?

Reply:
io::Error cannot represent full information about error. We have not only Errno as error, we have Error enum:
```rust
enum ThreadInitError {
    #[display(fmt = "Cannot get information about old signal stack: {}", _0)]
    OldSigStack(Errno),
    #[display(fmt = "Cannot mmap space for signal stack: {}", _0)]
    Mmap(Errno),
    #[display(fmt = "Cannot set new signal stack: {}", _0)]
    SigAltStack(Errno),
}
```
How can I transform it to io:Error without information lost?

Reply:
You can return `Result<(), ThreadInitError` then

Reply:
No, because it's only for unix errors

Reply:
Does this even matter in out case?

Reply:
We can make more suitable errors when the same support  will be done for windows (if it will be done). Currently I think it's not a good way to move unix special errors into lib.rs

Reply:
We don't even need to move, just change function signature

Reply:
No we need to make the same function for windows

Reply:
Why?

Reply:
Make a look at `windows.rs`

Reply:
It doesn't need to be the same in case of error. The only requirement is to be printable:
```
if let Err(err) = unsafe { sys::init_for_thread() } {
    log::debug!("Cannot initialize lazy-pages for thread: {}", err);
    ...
}
```

Reply:
Hmm may be,  ok will try it

Reply:
Hmm, no. In this case we should make this error as public, which is not currently suitable. And I think it's very strange solution